shremote.py

```diff
@@ -524,60 +543,41 @@ def __init__(self, program_name, cmd_cfg, index = None):
-                self.begin = float(begin_raw)
+            self.begin = try_eval(begin_raw, 'begin')
 
         self.index = index if index is not None else 0
         self.program = Program.get(program_name)
-
         self.log = Log.get(program_name)
         self.name = program_name
 
+        if self.program.start is None:
+            log_fatal("{}: Program requires a 'start' command".format(program_name))
+
         if 'duration' in cmd_cfg:
+            if self.program.stop is None:
+                log_fatal("{}: Must specify a 'stop' command if duration is specified".format(
+                          program_name))
             duration_raw = cmd_cfg.formatted('duration', **self.dict())
-            try:
-                dur = safe_eval(duration_raw, 'duration')
-                if str(dur) != str(duration_raw): # Means that 'eval' did something
-                    log_warn("Specifying evaluatable duration without $(...) is deprecated: {}".format(duration_raw))
-            except TypeError: ## This is fine, duration shouldn't have to be evaluated again
-                dur = duration_raw
-                pass
-
-            self.duration = float(dur)
-            if self.program.start is None:
-                log_fatal("{}: Cannot specify duration when program is missing 'start'".format(program_name))
-        elif self.program.start is not None and self.program.stop is not None:
-                log_fatal("{}: Must specify duration if program contains both 'start' and 'stop'".format(program_name))
-
-
-        if 'enforce_duration' in self.program.cfg:
-            enforce_duration = self.program.cfg.formatted('enforce_duration', **self.dict())
-            if enforce_duration == True:
-                if self.duration is not None:
-                    self.enforced_duration = self.duration - 1
-                else:
-                    self.enforced_duration = enforce_duration
-            else:
-                self.enforced_duration = self.program.cfg.enforce_duration
-        else:
-            if self.duration is not None:
-                self.enforced_duration = self.duration - 1
-            else:
-                self.enforced_duration = None
+            self.duration = int(try_eval(duration_raw, 'duration'))
 
+        #Enforce duration can be specified in the program or in the command
+        self.enforced_duration = None
+        enforce_duration_loc = None
         if 'enforce_duration' in cmd_cfg:
-            enforce_duration = cmd_cfg.formatted('enforce_duration', **self.dict())
+            enforce_duration_loc = cmd_cfg
+        elif 'enforce_duration' in self.program.cfg:
+            enforce_duration_loc = self.program.cfg
+
+        if enforce_duration_loc is not None:
+            enforce_duration = enforce_duration_loc.formatted('enforce_duration', **self.dict())
+            if not isinstance(enforce_duration, bool):
+                log_fatal("{}: enforce_duration must be a boolean".format(program_name))
```

Comment:
I left this as not-necessarily a boolean for a reason.
Sometimes there is no duration specified for a command, but there is an enforced duration. Such as with the runtimes for Dedos, where they are supposed to be stopped by the global controller stopping (so there's no specific `duration`), but you do want a duration to be enforced.
Other times there might be an upper-limit on how long a program should be run for, so it will be killed after that duration passes, but you want to make sure it runs for at least a lower-limit, which is the enforced duration